Rebase: Integrate uploader_optimization branch with update_agents and enhance agent/RAG handling

dagemawinegash opened the pull request:
## Description

### Overview
This PR rebases the `uploader_optimization` branch onto the `update_agents` branch to integrate the latest team changes, resolve conflicts, and incorporate the new agent system with RAG-specific enhancements.

### Changes Made

#### **Rebase Integration**
- **Rebased `uploader_optimization`** onto `update_agents` branch
- **Integrated team's latest changes** from the update_agents branch
- **Resolved merge conflicts** between the two branches

#### **Agent System & RAG Enhancements**
- **Integrated the updated agent orchestration** from the update_agents branch, replacing legacy AutoGen code
- **Customized agent logic for RAG-specific queries**, including improved handling of `pdf_ids` and context
- **Fixed duplicate user history records** by ensuring only one entry per query is created in the database
- **Added PDF summaries context to both classifier and LLM prompts** for improved query classification and enhanced conversation context
- **Implemented helper method for PDF summaries** and support filtering by `pdf_ids`

#### **Code Cleanup**
- **Fixed name mismatches** in various files
- **Cleaned up code structure** for better consistency
- **Ensured proper integration** with team's changes

#### **File Organization**
- **Separated SQL and Redis logic** into their specific dedicated files
- **Improved code organization** and maintainability
- **Maintained existing functionality** while integrating new structure

### Integration Details
- Successfully merged team's `update_agents` changes with local `uploader_optimization` work
- Preserved all existing functionality while incorporating new team features and agent improvements
- Ensured codebase consistency across the integrated branches

dagemawinegash committed:
Add PDF processing integration and vector-based search support

- Introduced PDF processing utilities for text extraction, cleaning, Markdown conversion, and content chunking.
- Extended main application routes with endpoints for PDF upload (`/upload_pdf`), question answering (`/ask_question`), and user status/data management.
- Implemented Qdrant-backed vector search and embedding classes to enable efficient document retrieval.
- Integrated LLM interface to process and answer queries.
- Developed JSON-based user data storage and update functions for managing PDF metadata.
- Added necessary dependencies

dagemawinegash committed:
Updated rag prompts to include PDF PROCESSOR PROMPT

dagemawinegash committed:
Implemented different features for better functionality
- Add /delete_pdf endpoint to remove specific PDF files by ID
- Integrate Qdrant vector DB deletion to purge associated embeddings
- Extend storage service to delete PDF files and clean up metadata entries
- Added extended metadata for the pdf including number of pages, file size, upload time and summary.
- Added feature that allow audio generation for query.

dagemawinegash committed:
Added the logic for extracting keywords, topics, summary with audio and suggested questions

dagemawinegash committed:
added the logic for creating audios for both the summary and query responses

dagemawinegash committed:
added the necessary prompts and dependencies

dagemawinegash committed:
refactor RAG and embedding logic, unify Qdrant and model usage
- Added the functionality of sentence transformers usage for embedding in llm_models.py
- bring all rag related logics to rag.py for uniform structure
- added new endpoints in routes.py for getting query and summary audio and for getting and cleaning user history
- unified functions that did similar task in to one in the qdrant.py

dagemawinegash committed:
Update PDF analyzer for unified RAG and summary storage

dagemawinegash committed:
Implement Redis audio caching and refactor TTS utils
- make changes in TTs utils for generating audio on demand
- added redis caching for caching audios

dagemawinegash committed:
Improve history and conversation prompt handling
- added a query id for a specific query and added a get_entry_by_query_id function to get a specific query so that later it can be retrived for audio generation

dagemawinegash committed:
Update app init and main for shared resources
- add a single reference of truth for qdrant client and other shared resources in the __init__.py
- updated main.py to use the shared resources from the init.py
- added history usage for context in main.py
- fixed a name mismatch occured in routes.py

dagemawinegash committed:
Refactor memory layer for consistency

dagemawinegash committed:
Remove unused llm_wrapper.py

dagemawinegash committed:
added user pdf file table and CRUD operation methods on this table.

dagemawinegash committed:
added history manager class to manage history related database queries
- added functions like create_history, retrieve_user_history, get_entry_by_query_id and clear_user_history for managing user history

dagemawinegash committed:
enforce form data for all endpoints and changed json queries to sql database queries

dagemawinegash committed:
refactor: update main and rag logic for new history and model structure

dagemawinegash committed:
added error handling log for storing general site information

dagemawinegash committed:
added useful note concerning embedding model choice

dagemawinegash committed:
added more relatable example with rejuve bio

dagemawinegash committed:
removed unused file

dagemawinegash committed:
added uncommited changes before rebasing

dagemawinegash committed:
added uncommited changes before rebasing

dagemawinegash committed:
code clean up and fixed name mismatch

dagemawinegash committed:
separated the logic of sql and redis to their specific files

dagemawinegash committed:
Integrate updated agent system from update-agents branch and tailored it for rag specific queries

kewserseid commented on app/main.py:
```diff
@@ -145,139 +176,181 @@ def get_galaxy_tools(query: str, user_id: str) -> str:
         - annotation: Requests for factual information about genes, proteins, variants,
         - hypothesis: Requests for Generation of a hypothesis graph on variant and phenotypes mentioned
         - galaxy: Requests about Galaxy web tools, workflows, or Galaxy platform capabilities
         - rag: General information requests
```
lets reprompt to add rag also to include pdf profiles other than general requests. maybe consider adding user uploaded pdf for better classification.

dagemawinegash replied:
Done.

kewserseid commented on app/storage/sql_storage.py:
```diff
@@ -204,20 +244,84 @@ def update_user_information(self, question_id: str, memory: dict = None, context
+        finally:
+            db.close()
+
+
```
Can we relate the tables using the user_id, so that when querying a user's history in the get_context_and_memory function, we can also retrieve the uploaded PDF associated with that user? This could be done by using a ForeignKey relationship. Additionally, this would allow us to include the uploaded PDF when querying in the assistant_response function inside main.py.

dagemawinegash replied:
I have made this as you said when we talked.

dagemawinegash committed:
Add PDF summaries to classifier and LLM prompts

- Include PDF summaries in classifier prompt for better query classification
- Pass PDF summaries to conversation prompt for enhanced LLM context
- Support filtering by pdf_ids when provided
- Update conversation prompt template to include PDF summaries context
